x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -511,7 +511,7 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
             // remove any already discovered UnresolvedAttributes that are in fact aliases defined later down in the tree
             // for example "from test | eval x = salary | stats max = max(x) by gender"
             // remove the UnresolvedAttribute "x", since that is an Alias defined in "eval"
-            AttributeSet planRefs = Expressions.references(p.expressions());
+            AttributeSet planRefs = p.references();
```

Comment:
Plans have logic to correctly calculate their references, no need to reinvent the wheel here.

In general, IMHO this method is trying to do too much.
The logic should be simple: each plan should declare needed inputs and produced outputs; this method should just start from the top of the plan, visit it down and for each plan node 
- remove the outputs 
- add the required inputs. 

All the `instanceof` above should be encapsulated in the single logical plan classes.

Reply:
Ideally, yes. I was about to suggest a refactor of `fieldNames` to make things simpler and more local, similarly to what you described.

However, since we're at a point here where references are not yet resolved, we cannot use the existing query plan methods as we would wish. Esp. the plan's `output` and `outputSet` methods _do not work_ here because these require resolved references, and also need to know about what fields are even present in relations. This, and the fact that we _need to match references by name_ rather than by _id_. This prevents us from just using the simple approach from [ProjectAwayColumns](https://github.com/elastic/elasticsearch/blob/a245e709ba5a94ad7a476a84d43f0b04bd361fc4/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/ProjectAwayColumns.java#L82-L84).

However, using the `references` method should be okay at this point because that's something that can be determined locally for a plan node.